### PR TITLE
check request range and fix response Content-Range.

### PR DESCRIPTION
**Pull request by sinnren**
Adjusted content 

The return value of the "apply_ranges" in the "write_response_core" is adjusted to bool. In the "apply_ranges", the validity of the requested range field is verified and saved to res.ranges_offset_length_ (new field) (using the original get_range_offset_and_length logic). Caching this fields are also needed for subsequent code and do not need to be processed again.

If "apply_ranges" returns false, it means that the range is wrong. Return 416 directly and clear the content.
The subsequent code only removes the offset verification, because all offsets and lengths have been verified in "apply_ranges".

Adjusted content in "make_content_range_header_field"
In the Content-Range field of the return value, if the request range-start or range-end is empty(-1),the offset and length of the actual returned content must be filled in, otherwise the browser will not work. (Use the Chrome view mp4 file, and the request is Range:1024-. If the return field does not fill in the range-end, such as "bytes 1024-/3124123", it cannot be played normally)

get_range_offset_and_length method is no longer needed, so removed

Several range processing logics are:
Assuming the content length is 10,
range:0-9 -> bytes 0-9/10, content-length:10
range:-3 -> bytes 7-9/10, content-length:3 (last 3 bytes)
range:-99 -> bytes 0-9/10, content-length:10 (fix length to 10, last 10 bytes)
range:0- -> bytes 0-9/10, content-length:10
range:0-99 -> bytes 0-9/10, content-length:10 (fix length to 10)
range:10- -> error:416 (range error)
range:10-99 -> error:416 (range error)

fix some string param(without .c_str())

ps.translated by google.

### CODE REVIEW

**Comment by yhirose** [comment]
@sinnren thanks for the pull request. Since it's a pretty big change, could you add enough unit tests in `test/test.cc`?

**Comment by sinnren** [comment]
> @sinnren thanks for the pull request. Since it's a pretty big change, could you add enough unit tests in `test/test.cc`?

16 lines added in test.cc, need more?

**Comment by yhirose** [comment]
@sinnren what kind of changes did you make in `test/test.cc`? It says 'Binary file'...

image

> 16 lines added in test.cc, need more?

I can't see your test code due to the above 'binary' file problem. Anyway, could you cover all the normal cases and edge cases?

**Comment by sinnren** [comment]
My mistake was that the encoding of the test.cc file was changed and turned into a binary file. It has been adjusted.

File: httplib.h [hunk 1]
```diff
@@ -454,11 +454,50 @@ class ContentReader {
...
+    return std::string();
+  }
+}
+
 using Range = std::pair<ssize_t, ssize_t>;
 using Ranges = std::vector<Range>;
 
 struct Request {
-  std::string method;
+  RequestMethod method = UNKNOWN;
```
Comment (yhirose) [inline httplib.h:500]: This is a breaking change and will cause build errors of existing users's code using `send` method. Could you change back `method` to `std::string` to `RequestMethod` please?

Comment (sinnren) [reply httplib.h:500]: oh ， I'll revert this change.

Comment (sinnren) [reply httplib.h]: reverted the breaking chage.

File: httplib.h [hunk 2]
```diff
@@ -524,8 +524,8 @@ struct Response {
   void set_header(const std::string &key, const std::string &val);
 
   void set_redirect(const std::string &url, int status = 302);
-  void set_content(const char *s, size_t n, const std::string &content_type);
-  void set_content(const std::string &s, const std::string &content_type);
+  void set_content(const char *s, size_t n, const std::string &content_type = {});
```
Comment (yhirose) [inline httplib.h:527]: Could you remove the default parameter `= {}` please?

File: httplib.h [hunk 3]
```diff
@@ -524,8 +524,8 @@ struct Response {
...
+  void set_content(const std::string &s, const std::string &content_type = {});
```
Comment (yhirose) [inline httplib.h:528]: Could you remove the default parameter `= {}` as well please?

File: httplib.h [hunk 4]
```diff
@@ -786,6 +787,7 @@ class Server {
   Server &set_file_extension_and_mimetype_mapping(const std::string &ext,
                                                   const std::string &mime);
   Server &set_default_file_mimetype(const std::string &mime);
+  Server &set_default_content_mimetype(const std::string &mime);
```
Comment (yhirose) [inline httplib.h:790]: Could you remove `set_default_content_mimetype` please? You can see the default with `set_default_headers`.

File: httplib.h (hunk 3 above)
Comment (yhirose) [reply httplib.h:528]: You don't have to use `set_content` method at all if you want to set the content text only. You can simply do it by `res.body = s` `with std::string` or `res.body.assign(s, n)` with `const char*`.

https://github.com/yhirose/cpp-httplib/blob/20a7f088cea4df658ed66b17153148b75e64881e/httplib.h#L5220

**Comment by sinnren** [comment]
copy that.

**Comment by yhirose** [comment]
@sinnren I made changes to remove unnecessary `.c_str()` calls on the master branch. So could you please rebase your code? Then, I think this pull request will end up including only changes directly related to the Content-Range. After that, I'll start review your pull request.

**Comment by sinnren** [comment]
> @sinnren I made changes to remove unnecessary `.c_str()` calls on the master branch. So could you please rebase your code? Then, I think this pull request will end up including only changes directly related to the Content-Range. After that, I'll start review your pull request.


ok.

File: test/test.cc [hunk 5]
```diff
@@ -2549,7 +2550,7 @@ TEST_F(ServerTest, StaticFileRanges) {
           .find(
               "multipart/byteranges; boundary=--cpp-httplib-multipart-data-") ==
       0);
-  EXPECT_EQ("265", res->get_header_value("Content-Length"));
+  EXPECT_EQ("266", res->get_header_value("Content-Length"));
```
Comment (yhirose) [inline test/test.cc:2553]: Why did you change the value to '266'?

Comment (sinnren) [reply test/test.cc:2553]: the request is cli_.Get("/dir/test.abcde", {{make_range_header({{1, 2}, {4, -1}})}});
this range {4, -1}

before:
ret range is "bytes 4-/5" , it was wrong because the range-end is empty(in method make_content_range_header_field).


after:
ret range is "bytes4-4/5",fix the range-end.(the purpose of the PR ) 

so.add 1 byte.

full body is:

`
----cpp-httplib-multipart-data-REkZlTo5uvw80k3K
Content-Type: text/abcde
Content-Range: bytes 1-2/5

bc
----cpp-httplib-multipart-data-REkZlTo5uvw80k3K
Content-Type: text/abcde
Content-Range: bytes 4-4/5

e
----cpp-httplib-multipart-data-REkZlTo5uvw80k3K--
`

**Comment by sinnren** [comment]
Any further adjustments needed?

**Comment by yhirose** [comment]
@sinnren thanks for your responses. I reviewed the changes, however, I am not comfortable with `ranges_offset_length_` in `Response`. So I resolved this issue in a simpler way at 420c975. Thank you anyway for finding this bug.
